server/controllers/things/build_controller.js

```diff
@@ -25,11 +25,20 @@ module.exports = function (typeName) {
       }
     },
     post: function (req, res) {
-      lib.create(req.body)
-      .then(res.json.bind(res))
-      .catch(error_.Handler(res))
+      if (req.files) {
+        lib.upload(req.body, req.files.mediaUpload) // code smell: hard coded input id, configurable, or per type?
```

Comment:
@almereyda 

Do I understand correctly that:

`req.body` would be the metadata of the media file and `req.files.mediaUpload` the binary contents of the actual asset? Are they both POSTed on the same API method call?

On https://hack.allmende.io/transformaps-20170926-development?view#api-calls-involved-on-uploading-a-media-file-and-associating-it-with-a-poi (and as currently implemented) I am envisioning to do this in 2 calls...

If my statement is correct, our implementations are currently dissaligned. Please confirm  or elaborate so we can align both implementations.

Reply:
From the sequence diagrams I learn you expect to be POSTing a BLOB to a separate `/blob` endpoint associated to any `media/:id` route. From this I like the idea to separate the creation of a `media` metadata container (thing) from its filling. Yet I suggest an upload can be triggered in presence of a `multipart` stream in any PUT action and thus more tightly couple this to a single medium's endpoint.

There may be the need for additional handling of the side effects, i.e. on other *things* such as `place`, where one might want to avoid allowing for binary uploads.

Reply:
@almereyda I have been studying your proposal of coupling the upload of the binary contents of the asset together with its metadata in a POST (creation) or PUT (update) method on `/media/` and came up with another argument to keep them decoupled.

As described on https://hack.allmende.io/transformaps-20170926-development?view#api-calls-involved-on-updating-a-poi-by-uploading-a-new-asset, updating a media file does not involve a PUT call to the `media` endpoint but a POST call to `media/{mediaId}/version} thus storing a new version object on the DB and not updating the media file's metadata.

I see 2 options at this point to achieve an implementation which makes sense to me:

1. To leave the mutipart upload of the asset decoupled from the creation (POST) or update of a certain media file

2. To delegate the responsibility of the creation of new versions of a media file (journaling) to the data service. This means that the editor would not make any POST calls to `media/{mediaId}/versions` as it is currently the case ( see https://github.com/TransforMap/transformap-editor/pull/42/files#diff-e2815ec3044d7d3fc5c27a5a3ed1b358R115). The editor would just make a PUT call to `media/{mediaId}` and the data service would then store a new version of it.